### ISSUE TITLE
EGL surface creation fails under under certain circumstances (and when using gpu VO)

### ISSUE THREAD

**Issue by moosotc** [mpv-player/mpv#7129]
### mpv version and platform
mpv 0.30.0-102-g8a0929973d Copyright © 2000-2019 mpv/MPlayer/mplayer2 projects
 built on Thu Nov  7 06:28:47 MSK 2019
ffmpeg library versions:
   libavutil       56.30.100 (runtime 56.35.101)
   libavcodec      58.53.101 (runtime 58.60.100)
   libavformat     58.28.102 (runtime 58.34.101)
   libswscale      5.4.101 (runtime 5.6.100)
   libavfilter     7.56.101 (runtime 7.66.100)
   libswresample   3.4.100 (runtime 3.6.100)
ffmpeg version: N-95660-gfc7b6d5574

Linux archlinux 5.3.8-arch1-1 #1 SMP PREEMPT @1572357769 x86_64 GNU/Linux

### Reproduction steps
#!/bin/sh
printf "P1\n1 1\n" >v1.xpm
printf "P1\n1 2\n" >v2.xpm
ffmpeg -loglevel 0 -y -f u8 -i /dev/zero -aframes 1 a.wav
mpv --no-config -v -v --log-file log v1.xpm a.wav v2.xpm a.wav v1.xpm

### Expected behavior
No errors

### Actual behavior
Failure to create EGL surface when trying to show the picture the second time

### Log file
http://ix.io/21bs

**Comment by moosotc** [comment]
This is likely Mesa related, fwiw the gpu is from NUC8I3BEH

- pacman -Qi mesa
Name            : mesa
Version         : 19.2.3-1

**Comment by wm4** [comment]
Hm nice reproduction. I've seen something like this before, but always suspected the problem in Mesa. Will take a closer look.

**Comment by wm4** [comment]
Hm eglCreateWindowSurface just fails with EGL_BAD_ALLOC. The only strange thing is that we create a EGL context/window successfully _without_ mapping the X11 window, and then  destroy it. When we create it again, it fails. The strange thing is that everything is destroyed and recreated, and fails the second time.

The normally works, e.g. hit "_" twice while playing a video.

Strong indication that it's Mesa.

**Comment by moosotc** [comment]
wm4 <notifications@github.com> writes:

> Hm eglCreateWindowSurface just fails with EGL_BAD_ALLOC. The only
> strange thing is that we create a EGL context/window successfully
> _without_ mapping the X11 window, and then destroy it. When we create
> it again, it fails. The strange thing is that everything is destroyed
> and recreated, and fails the second time.
>
> The normally works, e.g. hit "_" twice while playing a video.

I believe that it can be hit even with repeated _ (might take *many*
attempts) but it in't properly reproducible so was never reported.

Furthemore ometimes it doesn't just result in error but a segfault
inside egl (if my memory serves me).

>
> Strong indication that it's Mesa.

Most likely.

**Comment by wm4** [comment]
> I believe that it can be hit even with repeated _ (might take *many* attempts) but it in't properly reproducible so was never reported.

Possibly if you're fast enough that the window is never shown? Generally, mpv tries to show the window only once a video frame is available.

**Comment by moosotc** [comment]
wm4 <notifications@github.com> writes:

>> I believe that it can be hit even with repeated _ (might take *many*
> attempts) but it in't properly reproducible so was never reported.
>
> Possibly if you're fast enough that the window is never shown?
> Generally, mpv tries to show the window only once a video frame is
> available.

Uhm... not sure what you mean by that, the window was shown... Many
times, what you might mean is that one of the _'s is received at
inopportune moment, then - yeah that is a possibilty...

**Comment by moosotc** [comment]
moosotc@gmail.com writes:

> wm4 <notifications@github.com> writes:
>
>>> I believe that it can be hit even with repeated _ (might take *many*
>> attempts) but it in't properly reproducible so was never reported.
>>
>> Possibly if you're fast enough that the window is never shown?
>> Generally, mpv tries to show the window only once a video frame is
>> available.
>
> Uhm... not sure what you mean by that, the window was shown... Many
> times, what you might mean is that one of the _'s is received at
> inopportune moment, then - yeah that is a possibilty...

Just triggerred it by keeping the _ pressed:

- mpv -demuxer rawvideo /dev/zero
 (+) Video --vid=1 (rawvideo 1280x720 25.000fps)
VO: [gpu] 1280x720 yuv420p
Track switched:
     Video --vid=1 (rawvideo 1280x720 25.000fps)
video: no
Track switched:
 (+) Video --vid=1 (rawvideo 1280x720 25.000fps)
VO: [gpu] 1280x720 yuv420p
Track switched:
     Video --vid=1 (rawvideo 1280x720 25.000fps)
video: no
Track switched:
 (+) Video --vid=1 (rawvideo 1280x720 25.000fps)
VO: [gpu] 1280x720 yuv420p
Track switched:
     Video --vid=1 (rawvideo 1280x720 25.000fps)
video: no
Track switched:
 (+) Video --vid=1 (rawvideo 1280x720 25.000fps)
VO: [gpu] 1280x720 yuv420p
Track switched:
     Video --vid=1 (rawvideo 1280x720 25.000fps)
video: no
[vo/gpu/opengl] Could not create EGL surface!
[vo/gpu] Failed initializing any suitable GPU context!
Error opening/initializing the selected video_out (--vo) device.
Video: no video
Track switched:
     Video --vid=1 (rawvideo 1280x720 25.000fps)
video: no


Exiting... (Interrupted by error)

**Comment by moosotc** [comment]
moosotc@gmail.com writes:

> moosotc@gmail.com writes:
>
>> wm4 <notifications@github.com> writes:
>>
>>>> I believe that it can be hit even with repeated _ (might take *many*
>>> attempts) but it in't properly reproducible so was never reported.
>>>
>>> Possibly if you're fast enough that the window is never shown?
>>> Generally, mpv tries to show the window only once a video frame is
>>> available.
>>
>> Uhm... not sure what you mean by that, the window was shown... Many
>> times, what you might mean is that one of the _'s is received at
>> inopportune moment, then - yeah that is a possibilty...
>
> Just triggerred it by keeping the _ pressed:
>
> - mpv -demuxer rawvideo /dev/zero
>  (+) Video --vid=1 (rawvideo 1280x720 25.000fps)
> VO: [gpu] 1280x720 yuv420p
> Track switched:
>      Video --vid=1 (rawvideo 1280x720 25.000fps)

[..snip..]

Another point worth noting is that _reliable_ reproduction (here)
depends on dimensions of two videos being different.

**Comment by Th3Fanbus** [comment]
I can reproduce this on my machine using the reproduction steps: http://ix.io/21ol

This machine has an AMD Radeon HD 7950 GPU [1002:679a], so the issue is not Intel-specific.

### mpv version and platform
mpv 0.30.0 Copyright © 2000-2019 mpv/MPlayer/mplayer2 projects
 built on Thu Nov  7 22:14:58 UTC 2019
ffmpeg library versions:
   libavutil       56.31.100
   libavcodec      58.54.100
   libavformat     58.29.100
   libswscale      5.5.100
   libavfilter     7.57.100
   libswresample   3.5.100
ffmpeg version: n4.2.1

Linux archlinux 5.3.8-arch1-1 #1 SMP PREEMPT @1572357769 x86_64 GNU/Linux

pacman -Qi mesa
Name            : mesa
Version         : 19.2.2-1

**Comment by wm4** [comment]
Yeah, I'm on amdgpu and I'm seeing this too. Did anyone report this to Mesa yet?

**Comment by moosotc** [comment]
wm4 <notifications@github.com> writes:

> Yeah, I'm on amdgpu and I'm seeing this too. Did anyone report this to Mesa yet?

I have'nt. Lacking intimate familiarity with egl (or exactly how mpv does
things) Mesa3D bug report from me would be next to useless.

**Comment by barolo** [comment]
I've spent whole morning trying to figure out what's going on, I'm using amdgpu [ RAVEN APU ] too, for me it always fails to create egl context with : vo=gpu gpu-context=x11egl  hwdec=vaapi
mesa-19.3.0_rc3
```
[vo/gpu/opengl] EGL_VERSION=1.5
[vo/gpu/opengl] EGL_VENDOR=Mesa Project
[vo/gpu/opengl] EGL_CLIENT_APIS=OpenGL OpenGL_ES 
[vo/gpu/opengl] Trying to create Desktop OpenGL context.
[vo/gpu/opengl] Could not choose EGLConfig for Desktop OpenGL!
[vo/gpu/opengl] Trying to create GLES 3.x context.
[vo/gpu/opengl] Could not choose EGLConfig for GLES 3.x!
[vo/gpu/opengl] Trying to create GLES 2.x context.
[vo/gpu/opengl] Could not choose EGLConfig for GLES 2.x!
[vo/gpu/opengl] Could not create a GL context.
[vo/gpu/x11] uninit ...
```
Is it a mesa bug?

**Comment by wm4** [comment]
Unrelated mesa bug, eglgears supposedly fails too.

**Comment by barolo** [comment]
> Unrelated mesa bug, eglgears supposedly fails too.

Yup

"eglgears_x11 
EGL_VERSION = 1.5
EGLUT: failed to choose a config"

I just have hard time figuring out how it happened, I don't remember updating mesa but checking repo it looks like whole 19.x line was bumped and it's all affected... How annoying

**Comment by NiLuJe** [comment]
@barolo: Some distros (hi, Gentoo) build Mesa with "surfaceless" as the first platform.

Try setting `EGL_PLATFORM=x11` first.

(Having `EGL_LOG_LEVEL=debug` mildly helped illuminate the issue when I was wondering why the hell contetx creation failed ;)).

**Comment by barolo** [comment]
> @barolo: Some distros (hi, Gentoo [since [mesa 19.2](https://github.com/gentoo/gentoo/blob/2abc8d0bb878cfa6804571de88b8b0c28e79fb99/media-libs/mesa/mesa-19.2.4.ebuild#L359)]) build Mesa with `surfaceless` as the first platform. According to [docs](https://www.mesa3d.org/egl.html), that makes it the de-facto native/default platform.
> 
> Try enforcing `EGL_PLATFORM=x11` in your env instead (assuming this is actually on X11, and not Wayland).
> 
> (Having `EGL_LOG_LEVEL=debug` set mildly helped illuminate the issue when I was wondering why the hell context creation failed on `eglChooseConfig` ;)).

Yup, That did it. Thanks! I'd never figure it out....

**Comment by NiLuJe** [comment]
@barolo: I just switched to an amdgpu system, so I'm utterly unfamiliar with mesa & co. (i.e., I have no idea who's doing the right thing, or not).

**Comment by wm4** [comment]
I created a separate issue for the thing discussed in last couple of posts, since it's a different bug: #7154

**Comment by wm4** [comment]
Technically this was probably mpv's bug, but either way I'm very displeased with this EGL garbage.
